test/e2e/windows/cpu_limits.go

```diff
@@ -0,0 +1,138 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package windows
+
+import (
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2ekubelet "k8s.io/kubernetes/test/e2e/framework/kubelet"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	"time"
+
+	"github.com/onsi/ginkgo"
+)
+
+var _ = SIGDescribe("[Feature:Windows] Cpu Resources", func() {
+	f := framework.NewDefaultFramework("cpu-resources-test-windows")
+
+	// The Windows 'BusyBox' image is PowerShell plus a collection of scripts and utilities to mimic common busybox commands
+	powershellImage := imageutils.GetConfig(imageutils.BusyBox)
+
+	ginkgo.Context("Container limits", func() {
+		ginkgo.It("should not be exceeded after waiting 2 minutes", func() {
+			ginkgo.By("Creating one pod with limit set to '0.5'")
+			podsDecimal := newCPUBurnPods(1, powershellImage, "0.5", "1Gi")
+			f.PodClient().CreateBatch(podsDecimal)
+			ginkgo.By("Creating one pod with limit set to '500m'")
+			podsMilli := newCPUBurnPods(1, powershellImage, "500m", "1Gi")
+			f.PodClient().CreateBatch(podsMilli)
+			ginkgo.By("Waiting 2 minutes")
+			time.Sleep(2 * time.Minute)
+			ginkgo.By("Ensuring pods are still running")
+			var allPods [](*v1.Pod)
+			for _, p := range podsDecimal {
+				pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(p.Name,
+					metav1.GetOptions{})
+				framework.ExpectNoError(err, "Error retrieving pod")
+				framework.ExpectEqual(pod.Status.Phase, v1.PodRunning)
+				allPods = append(allPods, pod)
+			}
+			for _, p := range podsMilli {
+				pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(p.Name,
+					metav1.GetOptions{})
+				framework.ExpectNoError(err, "Error retrieving pod")
+				framework.ExpectEqual(pod.Status.Phase, v1.PodRunning)
+				allPods = append(allPods, pod)
+			}
+			ginkgo.By("Ensuring cpu doesn't exceed limit by >5%")
+			for _, p := range allPods {
+				ginkgo.By("Gathering node summary stats")
+				nodeStats, err := e2ekubelet.GetStatsSummary(f.ClientSet, p.Spec.NodeName)
+				framework.ExpectNoError(err, "Error grabbing node summary stats")
+				found := false
+				cpuUsage := float64(0)
+				for _, pod := range nodeStats.Pods {
+					if pod.PodRef.Name != p.Name || pod.PodRef.Namespace != p.Namespace {
+						continue
+					}
+					cpuUsage = float64(*pod.CPU.UsageNanoCores) * 1e-9
+					found = true
+					break
+				}
+				framework.ExpectEqual(found, true, "Found pod in stats summary")
+				framework.Logf("Pod %s usage: %v", p.Name, cpuUsage)
+				usageMatch := (.5 * 1.05) > cpuUsage
```

Comment:
better to also check cpuUsage is valid, like it is not always 0.

Reply:
Sure updating right now